juriscraper/pacer/http.py

```diff
@@ -0,0 +1,170 @@
+        :param json: json object to post
+        :param kwargs: assorted keyword arguments
+        :return: requests.Response
+        """
+        kwargs.setdefault('timeout', 300)
+
+        if data:
+            pacer_data = self._prepare_multipart_form_data(data)
+            return super(PacerSession, self).post(url, files=pacer_data, **kwargs)
+
+        return super(PacerSession, self).post(url, data=data, json=json, **kwargs)
+
+    @staticmethod
+    def _prepare_multipart_form_data(data):
+        """
+        Transforms a data dictionary into the multi-part form data that PACER
+        expects as the POST body
+        :param data: dict of data to transform
+        :return: dict with values wrapped into tuples like:(None, <value>)
+        """
+        output = dict()
+        for key in data:
+            output[key] = (None, data[key])
+        return output
+
+
+def _make_login_url(court_id):
+    """Make a login URL for a given court id."""
+    if court_id == 'psc':
+        # training account
+        return 'https://dcecf.psc.uscourts.gov/cgi-bin/login.pl'
+    return 'https://pacer.login.uscourts.gov/csologin/login.jsf?pscCourtId=%s' % court_id
+
+
+def login(court_id, username, password):
+    """
+    Log into a PACER jurisdiction via the main PACER portal which should set our global PacerSession
+
+    :param court_id: id of the court to authenticate with
+    :param username: PACER username
+    :param password: PACER password
+    :return: new PacerSession configured with PacerSession token in cookie
+    """
+    if username == 'tr1234':
+        return _login_training(court_id, username, password)
+
+    url = _make_login_url(court_id)
+    logger.info("Logging into: %s at %s" % (court_id, url))
+
+    login_session = requests.Session()
+    login_session.headers['User-Agent'] = 'Juriscraper'
+    login_session.verify=False
+
+    # initial GET to login page to get JSESSIONID
+    r = login_session.get(url, timeout=60)
+    if not r.status_code == 200:
+        msg = 'Could not navigate to PACER central login url: %s' % url
+        logger.error(msg)
+        raise BadLoginException(msg)
+
+    # with our JSESSIONID, try the login
+    login_data = {
+        'login': 'login',
+        'login:loginName': username,
+        'login:password': password,
+        'login:clientCode': '',
+        'login:fbtnLogin': '',
+        'javax.faces.ViewState': 'stateless'
+    }
+    r = login_session.post(url, timeout=60, data=login_data, allow_redirects=False)
+
+    if r.status_code == 302:
+        # we should be redirected on success with cookies!
+        if not login_session.cookies.get('PacerSession', None, '.uscourts.gov', '/'):
+            logger.error('Failed to get a PacerSession token!')
+            raise BadLoginException('Failed to get a PacerSession token!')
+    else:
+        msg = 'Unknown PACER login error: http status %s' % r.status_code
+        if 'Invalid ID or password' in r.text:
```

Comment:
That's what I was thinking, yeah. I guess it could subclass the BadLoginException? I just want to be able to know if the code screwed up or if it's bad credentials...ideally without sniffing an exception string.